File: Backend/controllers/resourceController.js
const Resource = require('../models/Resource');

exports.getAllResources = async (req, res) => {
    try {
        const resources = await Resource.find().sort({ createdAt: -1 });
        const mappedResources = resources.map(resource => ({
            ...resource.toObject(),
            id: resource._id.toString(),
            quantity_available: resource.quantityAvailable
        }));
        res.status(200).json(mappedResources);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

exports.getResourceById = async (req, res) => {
    try {
        const resource = await Resource.findById(req.params.id);
        if (!resource) {
            return res.status(404).json({ message: 'Resource not found' });
        }
        const mappedResource = {
            ...resource.toObject(),
            id: resource._id.toString(),
            quantity_available: resource.quantityAvailable
        };
        res.status(200).json(mappedResource);
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

exports.createResource = async (req, res) => {
    try {
        // Admin authorization check
        if (req.user.role !== 'admin') {
            return res.status(403).json({ message: 'Access denied: Admin role required' });
        }

        console.log('Creating resource:', req.body);
        const { name, description, category, quantity_available } = req.body;
        const resource = new Resource({
            name,
            description,
            category,
            quantityAvailable: quantity_available,
            createdBy: req.user.id
        });
        await resource.save();

        const mappedResource = {
            ...resource.toObject(),
            id: resource._id.toString(),
            quantity_available: resource.quantityAvailable
        };
        res.status(201).json(mappedResource);
    } catch (error) {
        console.error('Error creating resource:', error);
        res.status(500).json({ message: error.message });
    }
};

exports.updateResource = async (req, res) => {
    try {
        // Admin authorization check
        if (req.user.role !== 'admin') {
            return res.status(403).json({ message: 'Access denied: Admin role required' });
        }

        console.log('Updating resource:', req.params.id, req.body);
        const { name, description, category, quantity_available } = req.body;
        const resource = await Resource.findByIdAndUpdate(
            req.params.id,
            {
                name,
                description,
                category,
                quantityAvailable: quantity_available
            },
            { new: true }
        );
        if (!resource) {
            return res.status(404).json({ message: 'Resource not found' });
        }

        const mappedResource = {
            ...resource.toObject(),
            id: resource._id.toString(),
            quantity_available: resource.quantityAvailable
        };
        res.status(200).json(mappedResource);
    } catch (error) {
        console.error('Error updating resource:', error);
        res.status(500).json({ message: error.message });
    }
};

exports.deleteResource = async (req, res) => {
    try {
        // Admin authorization check
        if (req.user.role !== 'admin') {
            return res.status(403).json({ message: 'Access denied: Admin role required' });
        }

        const resource = await Resource.findByIdAndDelete(req.params.id);
        if (!resource) {
            return res.status(404).json({ message: 'Resource not found' });
        }
        res.status(200).json({ message: 'Resource deleted successfully' });
    } catch (error) {
        res.status(500).json({ message: error.message });
    }
};

File: Backend-Java/src/main/java/com/orrs/controller/ResourceController.java
package com.orrs.controller;

import com.orrs.model.Resource;
import com.orrs.model.User;
import com.orrs.repository.ResourceRepository;
import com.orrs.security.CustomUserDetails;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/resources")
public class ResourceController {

    private static final String SUPER_ADMIN_EMAIL = "adhithyanshanmugam@gmail.com";

    @Autowired
    private ResourceRepository resourceRepository;

    private boolean isAdmin(User user) {
        return user != null && ("admin".equalsIgnoreCase(user.getRole()) || SUPER_ADMIN_EMAIL.equalsIgnoreCase(user.getEmail()));
    }

    private ResponseEntity<?> accessDenied() {
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(Map.of("message", "Access denied: Admin role required"));
    }

    // GET /resources - Get all resources (public)
    @GetMapping
    public ResponseEntity<?> getAllResources() {
        List<Resource> resources = resourceRepository.findAllByOrderByCreatedAtDesc();
        return ResponseEntity.ok(resources);
    }

    // GET /resources/:id - Get resource by id (public)
    @GetMapping("/{id}")
    public ResponseEntity<?> getResourceById(@PathVariable("id") String id) {
        Optional<Resource> resourceOpt = resourceRepository.findById(id);
        if (resourceOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Resource not found"));
        }
        return ResponseEntity.ok(resourceOpt.get());
    }

    // POST /resources - Admin creates resource
    @PostMapping
    public ResponseEntity<?> createResource(@AuthenticationPrincipal CustomUserDetails userDetails, @RequestBody Resource request) {
        User currentUser = userDetails.getUser();
        if (!isAdmin(currentUser)) {
            return accessDenied();
        }

        if (request.getName() == null || request.getName().trim().isEmpty()) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(Map.of("message", "Resource name is required"));
        }

        Resource resource = new Resource();
        resource.setName(request.getName().trim());
        resource.setPrice(request.getPrice());
        resource.setCreatedBy(currentUser.getId());

        resourceRepository.save(resource);
        return ResponseEntity.status(HttpStatus.CREATED).body(resource);
    }

    // PUT /resources/:id - Admin updates resource
    @PutMapping("/{id}")
    public ResponseEntity<?> updateResource(@AuthenticationPrincipal CustomUserDetails userDetails, @PathVariable("id") String id, @RequestBody Resource request) {
        User currentUser = userDetails.getUser();
        if (!isAdmin(currentUser)) {
            return accessDenied();
        }

        Optional<Resource> resourceOpt = resourceRepository.findById(id);
        if (resourceOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Resource not found"));
        }

        Resource resource = resourceOpt.get();
        if (request.getName() != null) {
            resource.setName(request.getName().trim());
        }
        resource.setPrice(request.getPrice());

        resourceRepository.save(resource);
        return ResponseEntity.ok(resource);
    }

    // DELETE /resources/:id - Admin deletes resource
    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteResource(@AuthenticationPrincipal CustomUserDetails userDetails, @PathVariable("id") String id) {
        User currentUser = userDetails.getUser();
        if (!isAdmin(currentUser)) {
            return accessDenied();
        }

        Optional<Resource> resourceOpt = resourceRepository.findById(id);
        if (resourceOpt.isEmpty()) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("message", "Resource not found"));
        }

        resourceRepository.deleteById(id);
        return ResponseEntity.ok(Map.of("message", "Resource deleted successfully"));
    }
}
